active/003-seq-feature.md

```diff
@@ -0,0 +1,180 @@
+[70, 100)                                    {'name': 'cds1'}
+[50, 200)   {'db_xref': ('EcoGene:EG11555', 'GeneID:944757...
+[120, 150)     {'db_xref': 'EcoGene:EG11555', 'name': 'gen3'}
+[120, 150)                                   {'name': 'gen4'}
+   ```
+
+## Types of feature queries
+* query by location. `IntervalIndex` module implements `IntervalTree` for fast query by location. To get features covering a unit or a region, we can do this:
+
+``` python
+>>> interval_md.loc[30, ]
+          TYPE STRAND  LEFT RIGHT  \
+[0, 100)  gene      +  True  True
+
+                                                  ATTIBUTES
+[0, 100)  {'name': 'gen1', 'gene_synonym': 'ECK0007; JW0...
+>>> interval_md.loc[Interval(10, 30), ]
+          TYPE STRAND  LEFT RIGHT  \
+[0, 100)  gene      +  True  True
+[0, 30)    CDS      +  True  True
+
+                                                 ATTRIBUTES
+[0, 100)  {'name': 'gen1', 'gene_synonym': 'ECK0007; JW0...
+[0, 30)                                    {'name': 'cds1'}
+```
+
+* query by attributes. It can be used to fetch multiple features or a single feature. For example:
+
+``` python
+>>> interval_md[interval_md.apply(lambda row: row['ATTRIBUTES']['name']=='cds1', axis=1)]
+          TYPE STRAND  LEFT RIGHT        ATTRIBUTES
+[0, 30)    CDS      +  True  True  {'name': 'cds1'}
+[70, 100)  CDS      +  True  True  {'name': 'cds1'}
+
+```
+
+## Operations that need update in `interval_metadata`
+
+There are some methods in `Sequence` and its child classes that needs to operate on `interval_metadata` accordingly. For all the situations below, in case users only want to rc the sequence and do not intend to use `interval_metadata`, we should provide an option to skip changes and only return the sequence without `interval_metadata`.
+
+* Reverse operation. `INTERVAL` needs to be updated:
+``` python
+# original interval for a 9 nt gene
+>>> a = Interval(2, 11, closed='left')
+# after reverse complemenary of a seq of 12 nt, the interval should be
+>>> a_rc = 12 - a
+>>> a_rc
+Interval(1, 10, closed='left')
+```
+
+* Complementary operation. The `STRAND` needs to be switched.
+
+* Concatenation. The intervals of all the sequences, except for the 1st one, needs to be updated.
+
+``` python
+# interval of a feature in seq1 of 10 nt
+>>> a1 = Interval(2, 9, closed='left')
+# interval of a feature in seq2
+>>> a2 = Interval(1, 3, closed='left')
+# after concat, the two features will have the intervals of:
+>>> a1  # still the same
+Interval(2, 9, closed='left')
+>>> a2 + 10
+Interval(11, 13, closed='left')
+```
+
+* Slicing. The intervals needs to reflect the change of the sequence. Only keep the features that overlaps with sliced sequence, and need to update the coordiate as well. One caveat is that if a feature is cut in the middle of the slicing with missing left or right ends, the `LEFT` or `RIGHT` needs to be changed to `TRUE`
+
+# Alternatives
+
+Stumped here. Any ideas?
```

Comment:
Would it be worth creating a custom object to represent this idea? It could be backed by whatever (pandas, bx-python, etc..). One thing that strikes me about this model is a lot of it has to be convention, and there is a very natural split between the arbitrary user-defined schema (`ATTRIBUTES`) and the information the library needs to accurately manipulate a feature in context (`INTERVAL`, `STRAND`, etc...). It feels like the "right" way to do this might be to have an object with first-class understanding of things like non-contiguous intervals, fuzzy ends, and the strand type. Ideally this object would also permit querying the user-schema more concisely than a lambda'd-apply (which is typically not a very performant way to interact with a dataframe)

How did BioPython/PyCogent solve this problem? More importantly what kinds of operations did their models enable? Ideally we would want to avoid a world of `for` loops and property mining when querying sequence features.